File: src/main/java/com/angorasix/projects/core/presentation/router/ProjectRouter.java
package com.angorasix.projects.core.presentation.router;

import static org.springframework.web.reactive.function.server.RouterFunctions.route;

import com.angorasix.projects.core.presentation.handler.ProjectHandler;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;

/**
 * Router for all Project related endpoints.
 * 
 * @author rozagerardo
 *
 */
@Configuration
public class ProjectRouter {

  private final transient ProjectHandler handler;

  /**
   * Main ProjectRouter builder.
   * 
   * @param handler {@link ProjectHandler} that provides the functionality
   */
  public ProjectRouter(final ProjectHandler handler) {
    this.handler = handler;
  }

  /**
   * Main RouterFunction configuration for all endpoints related to Projects.
   * 
   * @return the {@link RouterFunction} with all the routes for Projects
   */
  @Bean
  public RouterFunction<ServerResponse> projectRouterFunction() {
    return route()
        .path("/projects",
            builder -> builder.GET("", handler::listProjects)
                .GET("/{id}", handler::getProject)
                .POST("", handler::createProject))
        .build();
  }

}

File: src/main/kotlin/com/angorasix/projects/core/presentation/router/ProjectRouter.kt
package com.angorasix.projects.core.presentation.router

import com.angorasix.commons.reactive.presentation.filter.extractRequestingContributor
import com.angorasix.projects.core.infrastructure.config.configurationproperty.api.ApiConfigs
import com.angorasix.projects.core.presentation.handler.ProjectHandler
import org.springframework.web.reactive.function.server.RouterFunction
import org.springframework.web.reactive.function.server.coRouter

/**
 * Router for all Project related endpoints.
 *
 * @author rozagerardo
 */
class ProjectRouter(
    private val handler: ProjectHandler,
    private val apiConfigs: ApiConfigs,
) {
    /**
     * Main RouterFunction configuration for all endpoints related to Projects.
     *
     * @return the [RouterFunction] with all the routes for Projects
     */
    fun projectRouterFunction() =
        coRouter {
            apiConfigs.basePaths.projectsCore.nest {
                filter { request, next ->
                    extractRequestingContributor(
                        request,
                        next,
                    )
                }
                apiConfigs.basePaths.baseByIdCrudRoute.nest {
                    path(apiConfigs.routes.validateAdminUser.path).nest {
                        method(apiConfigs.routes.validateAdminUser.method, handler::validateAdminUser)
                    }
                    method(apiConfigs.routes.getProject.method, handler::getProject)
                    method(apiConfigs.routes.updateProject.method, handler::updateProject)
                }
                method(apiConfigs.routes.listProjects.method, handler::listProjects)
                method(apiConfigs.routes.createProject.method, handler::createProject)
            }
        }
}
